python/dremio/middleware/auth_middleware.py

```diff
@@ -0,0 +1,59 @@
+"""
+  Copyright (C) 2017-2021 Dremio Corporation
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+"""
+
+from pyarrow import flight
+
+
+class DremioClientAuthMiddlewareFactory(flight.ClientMiddlewareFactory):
+    """A factory that creates DremioClientAuthMiddleware(s)."""
+
+    def __init__(self):
+        self.call_credential = []
+
+    def start_call(self, info):
+        return DremioClientAuthMiddleware(self)
+
+    def set_call_credential(self, call_credential):
+        self.call_credential = call_credential
+
+
+class DremioClientAuthMiddleware(flight.ClientMiddleware):
+    """
+    A ClientMiddleware that extracts the bearer token from
+    the authorization header returned by the Dremio
+    Flight Server Endpoint.
+    Parameters
+    ----------
+    factory : ClientHeaderAuthMiddlewareFactory
+        The factory to set call credentials if an
+        authorization header with bearer token is
+        returned by the Dremio server.
+    """
+
+    def __init__(self, factory):
+        self.factory = factory
+
+    def received_headers(self, headers):
+        auth_header_key = "authorization"
+        authorization_header = []
+        for key in headers:
+            if key.lower() == auth_header_key:
+                authorization_header = headers.get(auth_header_key)
```

Comment:
Same here. You are trying to get the value of the authorization header if it's present or raise an exception.
So it's headers -> header_pair? -> value | exception

This can be accomplished by filter and map or just by reduce.
```
authorization_header = reduce(lambda acc, header: header[1] if header[0] == auth_header_key else acc, headers.items())
```

Another issue here is that you break type consistency here — you initialize `authorization_header` as an Array but the you either get a String or exception. It becomes very apparent if you look at my implementation with reduce

Reply:
You make a great point here. The code also appears to try to encode the first letter(??) of the string to UTF-8 referencing authorization_header[0] instead of authorization_header.

To be fair, I didn't write this code, but should've caught this during my own review.